### ISSUE TITLE
Performance issue with the heatmap demo

### ISSUE THREAD

**Issue by spyhunter99** [osmdroid/osmdroid#820]
## Issue Type

[ ] Performance

## Description and/or steps/code to reproduce the problem

More samples > data > heatmap with async loading

once the "heat map" loads, which is just a set of non overlapping polygons with transparency, scroll panning is awful with hardware acceleration is on. With hardware acceleration off, it's smooth.

### If it's a bug, version(s) of android this affects:

testing on android 6

### Version of osmdroid the issue relates to:

6.0.0-SNAPSHOT on master

**Comment by monsieurtanuki** [comment]
I'll have a look at it.

**Comment by monsieurtanuki** [comment]
Here we are talking about hundreds of `Polygon`s being displayed!
I guess that for that demo (with square areas), simple `Path`s would be faster, because each time we `draw` a `Polygon` a lot of computation is involved.
What I can do is try to optimize the `LinearRing.buildPathPortion` part.
But the `HeatMap` demo itself could be optimized too.

**Comment by shiroyagi** [comment]
Yes, maybe manage the texture bitmap as mentioned on SO, or blend a hardware and a software layer if that is possible.

**Comment by monsieurtanuki** [comment]
I just had a go with a simple `drawRect` instead of a `Polygon`, and it looks very smooth to me.
Is there a specific reason why we should use `Polygon` for Heat Map, or are alternate solutions like `drawRect` acceptable? Something like that:
````java
new Overlay() {
	private final GeoPoint mNorthWest = new GeoPoint(key.getLatNorth(), key.getLonWest());
	private final GeoPoint mSouthEast = new GeoPoint(key.getLatSouth(), key.getLonEast());
	private final Point topLeft = new Point();
	private final Point bottomRight = new Point();

	@Override
	public void draw(Canvas c, MapView osmv, boolean shadow) {
		osmv.getProjection().toPixels(mNorthWest, topLeft);
		osmv.getProjection().toPixels(mSouthEast, bottomRight);
		c.drawRect(topLeft.x, topLeft.y, bottomRight.x, bottomRight.y, paint);
	}
};
````

**Comment by pasniak** [comment]
I have hardware acceleration on because otherwise map seems sluggish (on tablet).

Recently I added bunch of irregular map areas as Polygons (they are concave and most of the time they are outside the user's view) and when I switch on the layer, the map becomes painfully slow.

What are my options? Is it possible to disable HW acceleration on the fly?  I thought I had to set it before MapView is instantiated?

**Comment by monsieurtanuki** [comment]
@pasniak 
> I have hardware acceleration on because otherwise map seems sluggish (on tablet).
Recently I added bunch of irregular map areas as Polygons [...] and when I switch on the layer, the map becomes painfully slow.

`Polygon`s are implemented as `Path`s, which is notoriously slow when hardware acceleration is on.

> [My Polygons] are concave and most of the time they are outside the user's view

I'm going to create a dedicated issue. `Path`s being slow with hardware acceleration on is one thing, but when `Path`s are actually invisible they shouldn't be a source of low performance.

> Addition: my polygons are not filled in. I found filling in distracting. Maybe you could draw polygons as lines?

Use `Polyline` instead of `Polygon`:
* don't forget to add the initial point at the end - `Polygon`s are always closed, and `Polyline`s are never closed
* `Polyline`s use `Canvas.drawLines` instead of drawing a `Path`, and the former is much faster

**Comment by pasniak** [comment]
Sorry for confusion, when Polygons are totally outside of view, scrolling is indeed fast.   Having only a dozen of Polygons slows down panning visibly. Having just one Polygon in view slows rotation to crawl...  

I'd go with Polyline but how to I handle info touch inside?  I do have algo for detection of point inside a non-concave polygon but I mine are concave :( 

If no easy solution clicking on the edge would do - it is just less intuitive.

**Comment by monsieurtanuki** [comment]
@pasniak 
> when Polygons are totally outside of view, scrolling is indeed fast

I think it's worth optimizing anyway. That should be a good thing for your "touch inside" need, as well as for all Path displays that are partly out of the screen.

**Comment by pasniak** [comment]
I switched to Polylines. All is good except the labels on touch (on the edge) do not move when panning.  I think they move for Polygons.

**Comment by monsieurtanuki** [comment]
@spyhunter99 @pasniak Still relevant with the latest merged code?
